optimizer: supports callsite annotations of inlining, fixes #18773

aviatesk opened the pull request:
Enable `@inline`/`@noinline` annotations on function callsites.
From #40754.

Now `@inline` and `@noinline` can be applied to a code block and then
the compiler will try to (or try not to) inline calls within the block:
```julia
@inline f(...) # The compiler will try to inline `f`

@inline f(...) + g(...) # The compiler will try to inline `f`, `g` and `+`

@inline f(args...) = ... # Of course annotations on a definition is still allowed
```

Here are couple of notes on how those callsite annotations will work:
- callsite annotation always has the precedence over the annotation
  applied to the definition of the called method, whichever we use
  `@inline`/`@noinline`:
  ```julia
  @inline function explicit_inline(args...)
      # body
  end

  let
      @noinline explicit_inline(args...) # this call will not be inlined
  end
  ```
- when callsite annotations are nested, the innermost annotations has
  the precedence
  ```julia
  @noinline let a0, b0 = ...
      a = @inline f(a0)  # the compiler will try to inline this call
      b = notinlined(b0) # the compiler will NOT try to inline this call
      return a, b
  end
  ```
They're both tested and included in documentations.

Co-authored-by: Joseph Tan <jdtan638@gmail.com>

---

- closes #18773
- closes #40754

aviatesk committed:
optimizer: supports callsite annotations of inlining, fixes #18773

Enable `@inline`/`@noinline` annotations on function callsites.
From #40754.

Now `@inline` and `@noinline` can be applied to a code block and then
the compiler will try to (not) inline calls within the block:
```julia
@inline f(...) # The compiler will try to inline `f`

@inline f(...) + g(...) # The compiler will try to inline `f`, `g` and `+`

@inline f(args...) = ... # Of course annotations on a definition is still allowed
```

Here are couple of notes on how those callsite annotations will work:
- callsite annotation always has the precedence over the annotation
  applied to the definition of the called function, whichever we use
  `@inline`/`@noinline`:
  ```julia
  @inline function explicit_inline(args...)
      # body
  end

  let
      @noinline explicit_inline(args...) # this call will not be inlined
  end
  ```
- when callsite annotations are nested, the innermost annotations has
  the precedence
  ```julia
  @noinline let a0, b0 = ...
      a = @inline f(a0)  # the compiler will try to inline this call
      b = notinlined(b0) # the compiler will NOT try to inline this call
      return a, b
  end
  ```
They're both tested and included in documentations.

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -21,21 +21,30 @@ function push!(et::EdgeTracker, ci::CodeInstance)
+    elseif isa(src, OptimizationState) && isdefined(src, :ir)
+        return (is_stmt_inline(stmt_flag) || src.src.inlineable) ? src.ir : nothing
+    elseif src === nothing && is_stmt_inline(stmt_flag) && isa(match, MethodMatch)
+        # when the source isn't available at this moment, try to re-infer and inline it
+        # NOTE we can make inference try to keep the source if the call is going to be inlined,
+        # but then inlining will depend on local state of inference and so the first entry
+        # and the succeeding ones may generate different code; rather we always re-infer
+        # the source to avoid the problem while it's obviously not most efficient
+        # HACK disable inlining for the re-inference to avoid cycles by making sure the following inference never comes here again
+        interp = NativeInterpreter(get_world_counter(interp); opt_params = OptimizationParams(; inlining = false))
+        src, rt = typeinf_code(interp, match.method, match.spec_types, match.sparams, true)
+        return src
```
I tried the approach to use uncached source when the callsite is forced to be inlined (a1b1435f84f5891945dc233d73e606c69155098c), but it turns out that then inlining will depend o local state of inference and so the first entry and the succeeding ones may generate different code, e.g. from a test case I added:
```julia
function multilimited(a)
    if @noinline(isType(a))
        return @inline(multilimited(a.parameters[1]))
    else
        return rand(Bool) ? rand(a) : @inline(multilimited(a))
    end
end

# output might be different if we depend on inference local state
code_typed(m.multilimited, (Any,))
code_typed(m.multilimited, (Any,))
```

We might be able to keep source so that it doesn't depend on the local state, but it could be tricky.
Rather I'd like to always just re-infer when a source is unavailable, at least in this first iteration.

/cc @vtjnash

vtjnash replied:
I think this needs to happen during inference. It seems quite wrong to make a new NativeInterpreter here, and can probably have many other problems. We used to do some amount of recursion here, and we have gotten rid of it.

aviatesk replied:
What do you think about fea41f2654ac97d1ef59a7ac055c66e76b1bad57 ? We still need fresh optimization w/o inlining hack because those uncached sources are not optimized.

Well, now I'm wondering if it's really useful. I benchmarked some functions that hit this pass, and I couldn't find any cases where this "single-level-inlining" can help the performance.. Honestly I don't think it's worth the complexity that comes with this support.

aviatesk replied:
https://github.com/JuliaLang/julia/pull/41328#issuecomment-904460017

tisztamo commented:
Does callsite `@inline` force inlining completely, ignoring the cost model? (also close #40292?)

tisztamo commented:
Does callsite `@inline` force inlining completely, ignoring the cost model? (Also close #40292?) (re-commenting after accidental delete)

aviatesk commented:
> Does callsite @inline force inlining completely, ignoring the cost model? 

It will ignore the cost model, but not "force inlining completely", since there are some cases when a method body itself is unavailable (think when there are recursive calls but annotated as `@inline`).

I would appreciate if you could test this PR for your target, though.

tisztamo commented:
Just tested and it works, so this also closes #40292. Thank you very much!

vtjnash commented on base/compiler/optimize.jl:
```diff
@@ -374,16 +391,38 @@ function convert_to_ircode(ci::CodeInfo, code::Vector{Any}, coverage::Bool, sv::
+                end
+            end
         end
     end
```
I wonder if it would be better process all of these flags better during `jl_code_info_set_ir`, since we already pre-process a number of other flags there, as part of converting from unattached lowered form to associated source code form.

aviatesk replied:
Can we just pack it into `CodeInfo.ssaflags` ?

vtjnash replied:
Yes. We currently don't.

aviatesk replied:
what do you think on ff0b8a9cdb90568d34a1a42180bb6a5877e69b30 ?

aviatesk committed:
set ssaflags on `CodeInfo` construction

vtjnash commented on src/method.c:
```diff
@@ -297,10 +307,60 @@ static void jl_code_info_set_ir(jl_code_info_t *li, jl_expr_t *ir)
+                inbounds_depth = 0;
+            else if (inbounds_depth > 0)            // pop
+                inbounds_depth -= 1;
+        }
```
```suggestion
            bd[j] = jl_nothing;
        }
```
usually after processing metadata, we remove it

aviatesk replied:
I think it's still useful for debugging to keep them in lowered form ?

vtjnash commented on src/method.c:
```diff
@@ -297,10 +307,60 @@ static void jl_code_info_set_ir(jl_code_info_t *li, jl_expr_t *ir)
+                void* inline_flag = inline_flags->items[inline_flags->len - 1];
+                flag |= 1 << (inline_flag ? 1 : 2);
+            }
+            jl_arrayset(li->ssaflags, jl_box_uint8(flag), j);
```
```suggestion
            jl_array_uint8_set(li->ssaflags, j, flag);
```

aviatesk committed:
try to keep source if it will be force-inlined

aviatesk committed:
give inlining when source isn't available

aviatesk committed:
style nits

aviatesk commented:
Alright, I decided to give up the logic to keep otherwise-uncached source during abstractinterpret and re-optimize it to inline for now. To me it seems to be more trouble and complex than it's worth.
Still, with this PR we process statement flags before inference, and now  `const_prop_methodinstance_heuristic` respects statement-level `@inline` annotations, and so we can expect the annotations to work for most cases unless it's involved with recursive calls (I think the right solutions for them would be TCE and similar optimizations, rather than inlining).

So this PR ended up being quite simple at the end of the day, and I think it's ready to go.

aviatesk committed:
Update base/compiler/ssair/inlining.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

aviatesk committed:
Update src/method.c

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

aviatesk committed:
fixup

- remove preprocessed flags from `jl_code_info_set_ir`
- fix duplicated definition warning
- add and fix comments

aviatesk committed:
more clean up

aviatesk committed:
add caveat about the recursive call limitation

aviatesk committed:
update NEWS.md

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -21,23 +21,28 @@ function push!(et::EdgeTracker, ci::CodeInstance)
+        # (we can even natively try to re-infer it entirely)
+        # but it seems like that "single-level-inlining" is more trouble and complex than it's worth
+        # see https://github.com/JuliaLang/julia/pull/41328/commits/5557c2fe70d9672089a17c0f6a9f30fafcf0cb7c
+        return nothing
```
It's possible, but I'm afraid to spam with such warnings (they're really not interest from user's point of view, I think):
```julia
module Pkg
recursivef(...) = ...
function userfacingf(...)
  @inline recursivef(...)
end
export userfacingf
end

userfacingf(...) # we want warning ... ?
```

We can have it as opt-in feature, but that could be addressed in another PR.
Or, we can create new JET analyzer to report such cases.

aviatesk replied:
Okay, let's leave this to another PR.
Currently we don't have any agreement on how we want the compiler to emit warnings. E.g. it doesn't report anything back to us even if it finds obvious type errors during inference, but we could argue that it's more appropriate to get some warnings.

So I don't want this discussion to hold this PR.
If we adopt `@force_inline` naming, it should be done here before releasing it to nightly, but I'm afraid to accept it.
Given that there is no way to "force" inlining" with the existence of recursive calls, the essential difference between `@force_inline` and `@inline` is whether they emit the warning or not. And having two different names that does the same thing but the warning is more confusing than benefit imho.

aviatesk commented:
Rebased and added news entry for this.

Let me merge this tomorrow if there is no objection.

aviatesk commented:
It turns out this PR is very functional in actual execution, because inferred source is often discarded during codegen pipeline.
See https://github.com/JuliaLang/julia/issues/42078, and I may revert this PR for now.

KristofferC commented:
is -> isn't?

aviatesk commented:
ah, yeah thanks. silly typo...

aviatesk commented:
https://github.com/JuliaLang/julia/pull/42082 should fix the "serious" flaw with this PR, and now callsite inlining should work more reliably.

chriselrod commented:
@oscardssmith If you haven't tried this out yet:
```julia
julia> function vexp!(y,x)
           @inbounds for i ∈ eachindex(y,x)
               y[i] = exp(x[i])
           end
       end
vexp! (generic function with 1 method)

julia> function vexp_inline!(y,x)
           @inbounds for i ∈ eachindex(y,x)
               y[i] = @inline exp(x[i])
           end
       end
vexp_inline! (generic function with 1 method)

julia> x = randn(256); y = similar(x);

julia> @benchmark vexp!($y, $x)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.234 μs …  2.336 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.329 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.334 μs ± 49.344 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                       █▄                                    ▁
  ▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅██▁▇▃▁▆▁▁▁▁▁▁▁▁▁▁▁▃▁▁▁▁▃▁▁▁▁▁▆▃▄▄▇▆▅▆ █
  1.23 μs      Histogram: log(frequency) by time     1.49 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark vexp_inline!($y, $x)
BenchmarkTools.Trial: 10000 samples with 540 evaluations.
 Range (min … max):  212.893 ns … 316.361 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     213.315 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   214.200 ns ±   6.480 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▂▇█▇▃                   ▁▂▃▂▂▂▂                              ▂
  ▅█████▃▁▃▃▁▁▁▁▁▁▄▁▁▁▁▁▃▄▆████████▇██▇▇▇▆▅▄▃▅▄▅▃▁▄▄▃▄▄▄▄▄▅▃▄▄▅ █
  213 ns        Histogram: log(frequency) by time        220 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
